### ISSUE TITLE
Can't reproduce reported result on ImageNet

### ISSUE THREAD

**Issue by FreeZ3e** [EdwardChasel/Spatial-Mamba#27]
I trained a tiny version of Spatial Mamba on ImageNet, followed the reported training setting, and aligned the environment (Python 3.10, Torch 2.0.1, CUDA 11.8). 

However, I encounter the instability of grad_norm (inf or nan) in some EPOCHs, and the accuracy keeps going down after EPOCH 45.

Same in Python 3.10, Torch 2.3.0, CUDA 12.1

**Comment by WhoMeanJoy** [comment]
I have also encountered the same problem. Have you solved this problem? What should I do?

**Comment by FreeZ3e** [comment]
> I have also encountered the same problem. Have you solved this problem? What should I do?

In my environment, I found that this problem is caused by the lightweight depthwise convolution function:

    class StateFusion(nn.Module):
        def __init__(self, dim):
            super(StateFusion, self).__init__()
    
            self.dim = dim
            self.kernel_3   = nn.Parameter(torch.ones(dim, 1, 3, 3))
            self.kernel_3_1 = nn.Parameter(torch.ones(dim, 1, 3, 3))
            self.kernel_3_2 = nn.Parameter(torch.ones(dim, 1, 3, 3))
            self.alpha = nn.Parameter(torch.ones(3), requires_grad=True)
    
        @staticmethod
        def padding(input_tensor, padding):
            return torch.nn.functional.pad(input_tensor, padding, mode='replicate')
    
        def forward(self, h):
    
            if self.training:
                h1 = F.conv2d(self.padding(h, (1,1,1,1)), self.kernel_3,   padding=0, dilation=1, groups=self.dim)
                h2 = F.conv2d(self.padding(h, (3,3,3,3)), self.kernel_3_1, padding=0, dilation=3, groups=self.dim)
                h3 = F.conv2d(self.padding(h, (5,5,5,5)), self.kernel_3_2, padding=0, dilation=5, groups=self.dim)
                out = self.alpha[0]*h1 + self.alpha[1]*h2 + self.alpha[2]*h3
                return out
    
            else:
                if not hasattr(self, "_merge_weight"):
                    self._merge_weight = torch.zeros((self.dim, 1, 11, 11), device=h.device)
                    self._merge_weight[:, :, 4:7, 4:7] = self.alpha[0]*self.kernel_3
    
                    self._merge_weight[:, :, 2:3, 2:3] = self.alpha[1]*self.kernel_3_1[:,:,0:1,0:1]
                    self._merge_weight[:, :, 2:3, 5:6] = self.alpha[1]*self.kernel_3_1[:,:,0:1,1:2]
                    self._merge_weight[:, :, 2:3, 8:9] = self.alpha[1]*self.kernel_3_1[:,:,0:1,2:3]
                    self._merge_weight[:, :, 5:6, 2:3] = self.alpha[1]*self.kernel_3_1[:,:,1:2,0:1]
                    self._merge_weight[:, :, 5:6, 5:6] += self.alpha[1]*self.kernel_3_1[:,:,1:2,1:2]
                    self._merge_weight[:, :, 5:6, 8:9] = self.alpha[1]*self.kernel_3_1[:,:,1:2,2:3]
                    self._merge_weight[:, :, 8:9, 2:3] = self.alpha[1]*self.kernel_3_1[:,:,2:3,0:1]
                    self._merge_weight[:, :, 8:9, 5:6] = self.alpha[1]*self.kernel_3_1[:,:,2:3,1:2]
                    self._merge_weight[:, :, 8:9, 8:9] = self.alpha[1]*self.kernel_3_1[:,:,2:3,2:3]
    
                    self._merge_weight[:, :, 0:1, 0:1] = self.alpha[2]*self.kernel_3_2[:,:,0:1,0:1]
                    self._merge_weight[:, :, 0:1, 5:6] = self.alpha[2]*self.kernel_3_2[:,:,0:1,1:2]
                    self._merge_weight[:, :, 0:1, 10:11] = self.alpha[2]*self.kernel_3_2[:,:,0:1,2:3]
                    self._merge_weight[:, :, 5:6, 0:1] = self.alpha[2]*self.kernel_3_2[:,:,1:2,0:1]
                    self._merge_weight[:, :, 5:6, 5:6] += self.alpha[2]*self.kernel_3_2[:,:,1:2,1:2]
                    self._merge_weight[:, :, 5:6, 10:11] = self.alpha[2]*self.kernel_3_2[:,:,1:2,2:3]
                    self._merge_weight[:, :, 10:11, 0:1] = self.alpha[2]*self.kernel_3_2[:,:,2:3,0:1]
                    self._merge_weight[:, :, 10:11, 5:6] = self.alpha[2]*self.kernel_3_2[:,:,2:3,1:2]
                    self._merge_weight[:, :, 10:11, 10:11] = self.alpha[2]*self.kernel_3_2[:,:,2:3,2:3]
    
                out = DepthwiseFunction.apply(h, self._merge_weight, None, 11//2, 11//2, False)
    
                return out

So, I implement a vanilla version of this algorithm:

    class StateFusion(nn.Module):
        def __init__(self, dim):
            super(StateFusion, self).__init__()
            
            #self.dim = dim
            self.conv1 = nn.Conv2d(dim, dim, kernel_size=3, padding=1, dilation=1, groups=dim, bias=True)
            self.conv2 = nn.Conv2d(dim, dim, kernel_size=3, padding=3, dilation=3, groups=dim, bias=True)
            self.conv3 = nn.Conv2d(dim, dim, kernel_size=3, padding=5, dilation=5, groups=dim, bias=True)
    
            self.alpha = nn.Parameter(torch.ones(3), requires_grad=True)
    
        def forward(self, h):
            h1 = self.conv1(h)
            h2 = self.conv2(h)
            h3 = self.conv3(h)
            out = self.alpha[0]*h1 + self.alpha[1]*h2 + self.alpha[2]*h3
            return out

I use this implementation to reproduce the reported result on ImageNet successfully (Tiny).
However, the efficiency (throughput) of the model will be slightly reduced after the use of this version.

**Comment by FreeZ3e** [comment]
The author seems to have achieved the reported throughput of the Spatial-Mamba through the lightweight implementation of the algorithm. However, the top-1 accuracy on ImageNet is hard to reproduce(in my env and cluster). On this issue, I look forward to the author's reply.

**Comment by WhoMeanJoy** [comment]
> The author seems to have achieved the reported throughput of the Spatial-Mamba through the lightweight implementation of the algorithm. However, the top-1 accuracy on ImageNet is hard to reproduce(in my env and cluster). On this issue, I look forward to the author's reply.

Thank you very much for your reply, it has been very helpful to me.

**Comment by EdwardChasel** [comment]
Sorry for late reply, the grad_norm may appear inf or nan at some point, but this only happens occasionally in some samples and will not cause the model to crash. For training our models, our implementation is not much different from your vanilla version, except for the padding method and initialization. And for throughput, we use reparameterization techniques and optimized CUDA kernels to achieve that during inference. Please do not use the fused kernels during training as the current version does not support gradient backward.

**Comment by FreeZ3e** [comment]
> Sorry for late reply, the grad_norm may appear inf or nan at some point, but this only happens occasionally in some samples and will not cause the model to crash. For training our models, our implementation is not much different from your vanilla version, except for the padding method and initialization. And for throughput, we use reparameterization techniques and optimized CUDA kernels to achieve that during inference. Please do not use the fused kernels during training as the current version does not support gradient backward.

Thanks for ur reply.
Yep, the 'nan' and 'inf' are normal during training, that's why I feel confused about the accuracy decrease. No anomaly except 'nan' and 'inf, it's hard to debug, so I re-implemented the StateFusion and reached the reported performance, but lost some efficiency.
